AVRO-1887 Integrate Yetus and fix tests

Fokko opened the pull request:
All,

First PR that will kick off the tests using Apache Yetus. It is a crude implementation since it will only kick off the `./build.sh`, but this will be refined later on. 

The C++ and the interop tests are failing and are therefore disabled for now. 

cc @kojiromike

Cheers, Fokko

cutting commented on build.sh:
```diff
@@ -41,11 +41,13 @@ do
       (cd lang/c; ./build.sh test)
-      (cd lang/c++; ./build.sh test)
+      # The C++ tests are failing
+      # https://issues.apache.org/jira/projects/AVRO/issues/AVRO-2230
```
AVRO-2230 is now marked as fixed in Jira.

Fokko replied:
Still failing :'(
```
Scanning dependencies of target avrocpp_s
[  1%] Building CXX object CMakeFiles/avrocpp_s.dir/impl/Compiler.cc.o
/testptch/unknown/lang/c++/impl/Compiler.cc:174:15: warning: â€˜std::string avro::nameof(const NodePtr&)â€™ defined but not used [-Wunused-function]
 static string nameof(const NodePtr& n)
               ^
[  2%] Building CXX object CMakeFiles/avrocpp_s.dir/impl/Node.cc.o
[  4%] Building CXX object CMakeFiles/avrocpp_s.dir/impl/NodeImpl.cc.o
/testptch/unknown/lang/c++/impl/NodeImpl.cc: In function â€˜std::string avro::{anonymous}::escape(const string&)â€™:
/testptch/unknown/lang/c++/impl/NodeImpl.cc:31:8: warning: â€˜autoâ€™ changes meaning in C++11; please remove it [-Wc++0x-compat]
   for (auto c : unescaped) {
        ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc:31:13: error: â€˜câ€™ does not name a type
   for (auto c : unescaped) {
             ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc:67:3: error: expected â€˜;â€™ before â€˜returnâ€™
   return s;
   ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc:67:3: error: expected primary-expression before â€˜returnâ€™
/testptch/unknown/lang/c++/impl/NodeImpl.cc:67:3: error: expected â€˜;â€™ before â€˜returnâ€™
/testptch/unknown/lang/c++/impl/NodeImpl.cc:67:3: error: expected primary-expression before â€˜returnâ€™
/testptch/unknown/lang/c++/impl/NodeImpl.cc:67:3: error: expected â€˜)â€™ before â€˜returnâ€™
In file included from /usr/include/boost/assert.hpp:51:0,
                 from /usr/include/boost/format/parsing.hpp:21,
                 from /usr/include/boost/format.hpp:50,
                 from /testptch/unknown/lang/c++/api/Exception.hh:24,
                 from /testptch/unknown/lang/c++/api/Node.hh:28,
                 from /testptch/unknown/lang/c++/api/GenericDatum.hh:29,
                 from /testptch/unknown/lang/c++/api/NodeImpl.hh:23,
                 from /testptch/unknown/lang/c++/impl/NodeImpl.cc:21:
/testptch/unknown/lang/c++/impl/NodeImpl.cc: In member function â€˜virtual void avro::NodeRecord::printJson(std::ostream&, int) constâ€™:
/testptch/unknown/lang/c++/impl/NodeImpl.cc:247:59: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     assert(defaultValues.empty() || (defaultValues.size() == fields));
                                                           ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc: In member function â€˜virtual void avro::NodePrimitive::printDefaultToJson(const avro::GenericDatum&, std::ostream&, int) constâ€™:
/testptch/unknown/lang/c++/impl/NodeImpl.cc:292:13: error: â€˜to_stringâ€™ is not a member of â€˜stdâ€™
       os << std::to_string(g.value<float>());
             ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc:295:13: error: â€˜to_stringâ€™ is not a member of â€˜stdâ€™
       os << std::to_string(g.value<double>());
             ^
/testptch/unknown/lang/c++/impl/NodeImpl.cc: In function â€˜std::string avro::{anonymous}::escape(const string&)â€™:
/testptch/unknown/lang/c++/impl/NodeImpl.cc:68:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
CMakeFiles/avrocpp_s.dir/build.make:100: recipe for target 'CMakeFiles/avrocpp_s.dir/impl/NodeImpl.cc.o' failed
make[2]: *** [CMakeFiles/avrocpp_s.dir/impl/NodeImpl.cc.o] Error 1
make[1]: *** [CMakeFiles/avrocpp_s.dir/all] Error 2
CMakeFiles/Makefile2:425: recipe for target 'CMakeFiles/avrocpp_s.dir/all' failed
make: *** [all] Error 2
Makefile:147: recipe for target 'all' failed
```

Fokko replied:
I've created [a new ticket](https://issues.apache.org/jira/browse/AVRO-2249) on Jira, the tests are still failing :(

cutting commented on build.sh:
```diff
@@ -54,25 +56,26 @@ do
       # run interop rpc test
-      /bin/bash share/test/interop/bin/test_rpc_interop.sh
+      # This is not working for now
+      # /bin/bash share/test/interop/bin/test_rpc_interop.sh
```
Are all interop tests failing, or only a few?  If only a few, I'd rather disable just those.

cutting commented on lang/java/ipc/src/test/java/org/apache/avro/DataFileInteropTest.java:
```diff
@@ -40,27 +40,27 @@ public static void printDir() {
-      };
-    readFiles(provider);
-  }
+  // @Test
```
Why are these commented out?  Are they failing?  If they're no longer required then we should just delete them.

cutting commented on lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java:
```diff
@@ -541,13 +541,3 @@ public Player build() {
     }
   }
 }
-
```
Not sure it's worth cleaning up auto-generated files...

kojiromike commented on build.sh:
```diff
@@ -40,39 +40,43 @@ do
+
+      # create interop test data
+      mkdir -p build/interop/data
+      (cd lang/java/avro; mvn -B -P interop-data-generate generate-resources)
```
Shouldn't these be written as `cd ... &&` rather than `cd ...;`?

Fokko replied:
Good point

kojiromike commented on build.sh:
```diff
@@ -40,39 +40,43 @@ do
       (cd lang/c; ./build.sh test)
-      (cd lang/c++; ./build.sh test)
+      # https://issues.apache.org/jira/browse/AVRO-2249
+      # (cd lang/c++; ./build.sh test)
```
:(

Fokko replied:
These tests should be fixed, but I'm resolving some other stuff with the Java tests first.

kojiromike replied:
I've actually been thinking we should have two run modes for this:

1. Exit as soon as something fails
2. Keep going, and report all failures at the end.

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+function buildtest_usage
+{
+  yetus_add_option "--verbose=<true|false>" "print output to console (default: false)"
+}
```
Please consider the more portable function definition form

```bash
buildtest_usage() {
...
}
```

Fokko replied:
Thanks

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+{
+  local i
+
+  for i in "$@"; do
```
Is it important to have the form `--verbose=true` instead of just `--verbose`?

Fokko replied:
Good point

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+    big_console_header "Running ${file}"
+
+    #shellcheck disable=SC2001
+    sanitized_filename=$(echo "${file}" | sed -e 's,[/\.],-,g')
```
```bash
$ file=$file f=${file//[.\\\/]/-}
+ file='this.is\a/file.with\many/strange\characters/in.it'
+ f=this-is-a-file-with-many-strange-characters-in-it
```

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+
+    if [ "$VERBOSE" == "true" ]; then
+      # Write both to stdout and the file using tee
+      eval "cd ${BASEDIR}; ./${file} test" | tee -a "${PATCH_DIR}/build-${sanitized_filename}.txt"
```
Why `eval` and not a subshell?

Fokko replied:
Good point

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+    #shellcheck disable=SC2001
+    sanitized_filename=$(echo "${file}" | sed -e 's,[/\.],-,g')
+
+    if [ "$VERBOSE" == "true" ]; then
```
```bash
if [[ $VERBOSE = true ]]; then
```

but it would be better, in my opinion, to consider VERBOSE as true if it's set at all.

kojiromike commented on share/precommit/buildtest.sh:
```diff
@@ -0,0 +1,81 @@
+      result=$?
+    fi
+
+    if [[ ${result} -ne 0 ]]; then
```
Consider arithmetic expression here:

```bash
if (( result != 0 )); then
```

Fokko replied:
Thanks

kojiromike commented on share/test/interop/bin/test_rpc_interop.sh:
```diff
@@ -21,29 +21,33 @@ cd `dirname "$0"`/../../../..   # connect to root
 
 VERSION=`cat share/VERSION.txt`
```
```bash
VERSION=$(<share/VERSION.txt)
```

would be a little nicer here. (Obviously take it or leave it since this isn't actually part of this PR.)

kojiromike commented on share/test/interop/bin/test_rpc_interop.sh:
```diff
@@ -21,29 +21,33 @@ cd `dirname "$0"`/../../../..   # connect to root
 portfile=/tmp/interop_$$
 
 function cleanup() {
   rm -rf $portfile
```
Please quote this expansion. An unquoted expansion after `rm -rf` is dangerous.

Fokko commented:
Thanks for the review @kojiromike 

Enabling `shellcheck` is easy by adding it in the list of plugins: https://github.com/Fokko/avro/blob/fd-fix-tests/.travis.yml#L34

There is also support for hadolint etc.

Fokko commented:
@kojiromike @cutting PTAL.

I know there are a crazy amount of changes. Instead of having fixed paths, I updated all the tests to use TemporaryDirectories using junit. Also, I've rewritten a lot of the file operations using `try-with-resource`. A lot of the streams weren't closed at all.

Fokko commented:
@kojiromike My suggestion would be to merge this first, and then you can freely rework the shell scripts. You're much better at it than me :-)

kojiromike commented:
@Fokko Congratulations/Thanks for getting this merged. I'll take a look at those shell scripts. Am I correct that now we can rely on automated checks to validate a lot of our PRs? Is there a way to trigger the tests on all the open PRs?

Fokko commented:
@kojiromike with pleasure. They need to rebase onto master, then Travis will run automatically. I'm going through a couple of the PR's that are still open.

aw-was-here commented:
FWIW, very specific support for Travis (and other CI systems) is something that will hopefully be coming to Yetus soon.   I've got it in a fork already: https://github.com/aw-was-here/buretoolbox/commit/f055aa3c78ffdd9bf42e11b457b1283bc3582960
